packages/iov-core/src/jsrpc.ts

```diff
@@ -0,0 +1,242 @@
+import {
+  jsonRpcCodeInternalError,
+  jsonRpcCodeInvalidParams,
+  jsonRpcCodeInvalidRequest,
+  jsonRpcCodeMethodNotFound,
+  jsonRpcCodeParseError,
+  jsonRpcCodeServerErrorDefault,
+  SimpleMessagingConnection,
+} from "@iov/jsonrpc";
+import { firstEvent } from "@iov/stream";
+
+// An RPC interface very similar to JSON-RPC but supporting
+// all the data types from the structured clone algorithm
+// https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#Supported_types
+
+/** A single structured clone algorithm compatible value */
+export declare type JsRpcCompatibleValue =
+  | string
+  | number
+  | boolean
+  | null
+  | Uint8Array
+  | JsRpcCompatibleArray
+  | JsRpcCompatibleDictionary;
+
+/** An array of JsRpcCompatibleValue */
+export interface JsRpcCompatibleArray extends ReadonlyArray<JsRpcCompatibleValue> {}
+
+/** A string to JsRpcCompatibleValue dictionary. */
+export interface JsRpcCompatibleDictionary {
+  readonly [key: string]: JsRpcCompatibleValue;
+}
+
+export function isJsRpcCompatibleValue(value: unknown): value is JsRpcCompatibleValue {
+  return (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean" ||
+    value === null ||
+    value instanceof Uint8Array ||
+    isJsRpcCompatibleArray(value) ||
+    isJsRpcCompatibleDictionary(value)
+  );
+}
+
+export function isJsRpcCompatibleArray(value: unknown): value is JsRpcCompatibleArray {
+  if (!Array.isArray(value)) {
+    return false;
+  }
+  return value.every(isJsRpcCompatibleValue);
+}
+
+export function isJsRpcCompatibleDictionary(data: unknown): data is JsRpcCompatibleDictionary {
+  if (typeof data !== "object" || data === null) {
+    // data must be a non-null object
+    return false;
+  }
+
+  // Exclude special kind of objects like Array, Date or Uint8Array
+  // Object.prototype.toString() returns a specified value:
+  // http://www.ecma-international.org/ecma-262/7.0/index.html#sec-object.prototype.tostring
+  if (Object.prototype.toString.call(data) !== "[object Object]") {
+    return false;
+  }
+
+  for (const key of Object.getOwnPropertyNames(data)) {
```

Comment:
looks good. but what about

`return Object.values(data).every(isJsRpcCompatibleValue)`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values

Reply:
Great find. This is exactly what the code tries to say. However, `Object.values` is only available in the ES2017 standard library and does not compile with our setup:

```
src/jsrpc.ts:66:17 - error TS2339: Property 'values' does not exist on type 'ObjectConstructor'.

66   return Object.values(data).every(isJsRpcCompatibleValue);
                   ~~~~~~


Found 2 errors.
```

Reply:
But this works

```ts
  // replace with Object.values when available (ES2017+)
  const values = Object.getOwnPropertyNames(data).map(key => (data as any)[key]);
  return values.every(isJsonCompatibleValue);
```